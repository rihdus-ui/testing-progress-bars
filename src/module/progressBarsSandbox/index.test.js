/**
 * Created by t_mehes on 3/7/17.
 */

import React from 'react';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import ReactTestUtils from 'react-dom/test-utils'
import axios from 'axios';
import sinon from 'sinon';

import ProgressbarView from './index.js';

describe('ProgressbarView', () => {
  let progressBarView = null;
  let sandbox = null
    , server = null
  ;
  beforeAll(() => {
    const data = {
      buttons: [31, 22, -48, -25],
      bars: [78, 71, 85, 73],
      limit: 200
    };
    progressBarView = renderer.create((<ProgressbarView
      maxVal={data.limit}
      selectBarId={0}
      bars={data.bars}
      buttons={data.buttons}
      loading={false}/>))
  });

  afterEach(() => {
  });

  it('component should render', () => {
    let tree = progressBarView.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('component render', () => {
    const data = {
      buttons: [31, 22, -48, -25],
      bars: [78, 71, 85, 73],
      limit: 200
    };
    let progressBarDataView = null
      , emptyView = null;

    beforeAll(() => {
      progressBarDataView = shallow((<ProgressbarView
        maxVal={data.limit}
        selectBarId={0}
        bars={data.bars}
        buttons={data.buttons}
        loading={false}/>));

      emptyView = shallow((<ProgressbarView
        maxVal={data.limit}
        selectBarId={0}
        bars={data.bars}
        buttons={data.buttons}
        loading={true}/>))
    });

    it('should render loading view but not the data view when data is being loaded', () => {
      const dataView = emptyView.find('.test-class-data-view');
      const loadingView = emptyView.find('.test-class-loading-view');
      expect(dataView.node).toBe(undefined);
      expect(loadingView.node).not.toBe(undefined);
    });

    it('should render data view but not the loading view when data loading flag is false', () => {
      const dataView = progressBarDataView.find('.test-class-data-view');
      const loadingView = progressBarDataView.find('.test-class-loading-view');
      expect(dataView.node).not.toBe(undefined);
      expect(loadingView.node).toBe(undefined);
    });

    it('should update progressbar selection on progress select change', () => {
      const dataView = progressBarDataView.find('.test-class-data-view');
      expect(dataView).not.toBe(null);
      let sandboxView = ReactTestUtils.renderIntoDocument(<ProgressbarView
        maxVal={data.limit}
        selectBarId={0}
        bars={data.bars}
        buttons={data.buttons}
        loading={false}/>);

      /** Test State update **/
      expect(sandboxView.state.selectBarId).toBe(0);
      sandboxView.onProgressSelect(1);
      expect(sandboxView.state.selectBarId).toBe(1);
    });


    it('should update progressbar value on update', () => {
      const dataView = progressBarDataView.find('.test-class-data-view');
      let sandboxView = ReactTestUtils.renderIntoDocument(<ProgressbarView
        maxVal={data.limit}
        selectBarId={0}
        bars={data.bars}
        buttons={data.buttons}
        loading={false}/>);

      /** Test State update **/
      const originalVal = data.bars[0];

      expect(sandboxView.state.selectBarId).toBe(0);
      expect(sandboxView.state.bars[0]).toBe(originalVal);

      /** Simple Increment */
      sandboxView.onProgressValueUpdate(10);
      expect(sandboxView.state.bars[0]).toBe(originalVal + 10);

      /** Simple Decrement */
      sandboxView.onProgressValueUpdate(-9);
      expect(sandboxView.state.bars[0]).toBe(originalVal + 10 - 9);


      /** Test value overflow
       * --------------------
       * Value should not exceed max value.
       * **/
      sandboxView.onProgressValueUpdate(1000);
      expect(sandboxView.state.bars[0]).toBe(data.limit);


      /** Test value underflow
       * ----------------------
       * Value should not be less than 0 if the user tries to reduce the
       * value below the min value.
       * **/
      sandboxView.onProgressValueUpdate(-10000);
      expect(sandboxView.state.bars[0]).toBe(0);
    })

  });

});