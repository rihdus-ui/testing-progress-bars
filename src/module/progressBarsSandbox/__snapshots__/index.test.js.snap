// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressbarView component should render 1`] = `
<div
  className="flexbox f-col flex1"
>
  <div
    className="test-class-data-view"
  >
    <div
      className="mb2 flexbox f-row text-class-progressbar-container"
    >
      <div
        className="p1 bar-prefix active"
      >
        #
        0
      </div>
      <div
        className="progressBar"
      >
        <div
          className="bar-container"
        >
          <svg
            className="bar-svg"
            height="100%"
            preserveAspectRatio="none"
            viewBox="0 0 100 10"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              className="path "
              strokeDasharray="100"
              strokeDashoffset={0}
              strokeWidth="10"
              style={
                Object {
                  "transform": "translateX(-61%)",
                }
              }
              x1="0"
              x2="100%"
              y1="5"
              y2="5"
            />
          </svg>
        </div>
        <div
          className="bar-text"
        >
          <span>
            39
          </span>
          <span>
            %
          </span>
        </div>
      </div>
    </div>
    <div
      className="mb2 flexbox f-row text-class-progressbar-container"
    >
      <div
        className="p1 bar-prefix "
      >
        #
        1
      </div>
      <div
        className="progressBar"
      >
        <div
          className="bar-container"
        >
          <svg
            className="bar-svg"
            height="100%"
            preserveAspectRatio="none"
            viewBox="0 0 100 10"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              className="path "
              strokeDasharray="100"
              strokeDashoffset={0}
              strokeWidth="10"
              style={
                Object {
                  "transform": "translateX(-64.5%)",
                }
              }
              x1="0"
              x2="100%"
              y1="5"
              y2="5"
            />
          </svg>
        </div>
        <div
          className="bar-text"
        >
          <span>
            36
          </span>
          <span>
            %
          </span>
        </div>
      </div>
    </div>
    <div
      className="mb2 flexbox f-row text-class-progressbar-container"
    >
      <div
        className="p1 bar-prefix "
      >
        #
        2
      </div>
      <div
        className="progressBar"
      >
        <div
          className="bar-container"
        >
          <svg
            className="bar-svg"
            height="100%"
            preserveAspectRatio="none"
            viewBox="0 0 100 10"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              className="path "
              strokeDasharray="100"
              strokeDashoffset={0}
              strokeWidth="10"
              style={
                Object {
                  "transform": "translateX(-57.5%)",
                }
              }
              x1="0"
              x2="100%"
              y1="5"
              y2="5"
            />
          </svg>
        </div>
        <div
          className="bar-text"
        >
          <span>
            43
          </span>
          <span>
            %
          </span>
        </div>
      </div>
    </div>
    <div
      className="mb2 flexbox f-row text-class-progressbar-container"
    >
      <div
        className="p1 bar-prefix "
      >
        #
        3
      </div>
      <div
        className="progressBar"
      >
        <div
          className="bar-container"
        >
          <svg
            className="bar-svg"
            height="100%"
            preserveAspectRatio="none"
            viewBox="0 0 100 10"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              className="path "
              strokeDasharray="100"
              strokeDashoffset={0}
              strokeWidth="10"
              style={
                Object {
                  "transform": "translateX(-63.5%)",
                }
              }
              x1="0"
              x2="100%"
              y1="5"
              y2="5"
            />
          </svg>
        </div>
        <div
          className="bar-text"
        >
          <span>
            37
          </span>
          <span>
            %
          </span>
        </div>
      </div>
    </div>
    <div
      className="flexbox f-row flex-wrap"
    >
      <div
        className="mr2 mt2"
      >
        <div
          className="select-container"
        >
          <select
            className="selector"
            id="bar-select"
            name="selectBar"
            onChange={[Function]}
          >
            <option
              value={0}
            >
              Progress #
              0
            </option>
            <option
              value={1}
            >
              Progress #
              1
            </option>
            <option
              value={2}
            >
              Progress #
              2
            </option>
            <option
              value={3}
            >
              Progress #
              3
            </option>
          </select>
        </div>
      </div>
      <span
        className="flex1"
      />
      <div
        className="ib"
      >
        <button
          className="button mt2 "
          max={200}
          onClick={[Function]}
          value={31}
        >
          31
        </button>
        <button
          className="button mt2 ml1"
          max={200}
          onClick={[Function]}
          value={22}
        >
          22
        </button>
        <button
          className="button mt2 ml1"
          max={200}
          onClick={[Function]}
          value={-48}
        >
          -48
        </button>
        <button
          className="button mt2 ml1"
          max={200}
          onClick={[Function]}
          value={-25}
        >
          -25
        </button>
      </div>
    </div>
  </div>
</div>
`;
